Mathlib/Data/Multiset/Powerset.lean

```diff
@@ -23,15 +23,21 @@ variable {α : Type*}
 
 /-! ### powerset -/
 
---Porting note: TODO: Write a more efficient version
```

Comment:
In what sense is the new version more efficient? In terms of algorithmic complexity they're the same, right? Is the problem the intermediate memory?

Reply:
If nothing else, it fixes timeouts and seems to be a more faithful port of the Lean 3 version...

Reply:
I suspect the cause of the slowdown was the use of `Array` in `List.sublists`. An `Array` in the Kernel is just a `List`, but compiled to something more efficient. `List.sublists` uses `Array.push`, which is slow in the kernel becuase it's `l ++ [a]` on the underlying list, so it's linear time instead of constant time. So this new implementation is a better complexity in the kernel.

Reply:
Does that mean `List.sublists` also is inefficient in the kernel? Can we change it back to be array-free, and then this definition will get the same improvement?

Reply:
If we're worried about runtime performance, then we can add a `csimp` lemma in Std that converts the kernel-friendly version into the execution-friendly version. The advantage of doing that in Std to `List.sublists` is that it should cause everything downstream to be optimal automatically.

Reply:
> If we're worried about runtime performance, then we can add a `csimp` lemma in Std that converts the kernel-friendly version into the execution-friendly version. The advantage of doing that in Std to `List.sublists` is that it should cause everything downstream to be optimal automatically.

I think this is a good suggestion for Std and `sublists`. Even if we do that, though, using `List.sublists` to generate a `List (List a)` and then mapping Multiset.ofList to obtain a `List (Multiset a)` is naturally more expensive than constructing a `List (Multiset a)` directly because there is more deduplication work to be performed. My proposal would be to leave this to future work.

Reply:
What deduplication are you thinking of?

Reply:
Sorry, you're right! I was confused by the appearance of `pwFilter` in traces, but that's only because these are later converted to Finsets.

Reply:
With that correction, do you agree that this PR probably isn't needed and we should just fix Std instead?

Reply:
I do agree it is plausible that changing `sublists` is enough, but my motivation for working on this was Bhavik mentioning on Zulip that this was blocking his full migration to Lean 4, and it's unclear if the extra overhead of using `sublists` and coercing from List to Multiset would matter for the application discussed on Zulip; the coercion is extra work for kernel reduction even if no deduplication is involved. I'm (unfortunately) also over my Lean time budget, so someone else would have to do the Std4 work if the intention is to get it merged soon. But I wouldn't mind if someone else did it, of course.

Reply:
For reference, my check for whether alternate definitions are "good enough" is the following example, which essentially comes from the Zulip thread:

```lean
import Mathlib.Order.Partition.Finpartition

open Finset

instance Finpartition.fintype_finset
  {α : Type _} [DecidableEq α] (a : Finset α) : Fintype (Finpartition a) where
  elems := a.powerset.powerset.image
    (λ p => if h : p.SupIndep id ∧ p.sup id = a ∧ ⊥ ∉ p then ⟨p, h.1, h.2.1, h.2.2⟩ else ⊥)
  complete := by
    rintro p
    rw [mem_image]
    refine' ⟨p.parts, _, _⟩
    · simp only [mem_powerset]
      intros i hi
      rw [mem_powerset]
      exact p.le hi
    · rw [dif_pos]
      simp only [p.supIndep, p.supParts, p.not_bot_mem, eq_self_iff_true, not_false_iff, and_self]

example : @Fintype.card (Finpartition (range 3)) (Finpartition.fintype_finset _) = 5 := by rfl
```

Reply:
@eric-wieser Answering your question again, I checked by reverting my changes and replacing the Std definition by 
```
  l.foldr (fun a acc => join (acc.map fun x => [x, a :: x])) [[]]
```
(and sorrying out the affected theorems) and the example in the previous comment didn't time out. Therefore, I fully agree that we should fix it in Std. Unfortunately I don't have the time to prepare and shepherd a Std PR right now, but I'd be happy if you or someone else did it.

Reply:
I've tried to adopt this in #7746

Reply:
Do recent elaborator changes have any effect on this problem?

Reply:
The performance difference is still noticable; see the example in the std issue.